EventGenerator/src/RPCGun_module.cc

```diff
@@ -94,176 +101,140 @@ namespace mu2e {
-    double _mbtime = accPar->deBuncherPeriod;
-
-    IO::StoppedParticleTauNormF stop;
-    if (tmin_ > 0){
-      while (true){
-        const auto& tstop = stops_.fire();
-        timemap->SetTime(protonPulse_->fire());
-        if (tstop.t+timemap->time() < 0 || tstop.t+timemap->time() > tmin_){
-          if (applySurvivalProbability_){
-            double weight = exp(-tstop.tauNormalized)*survivalProbScaling_;
-            if (weight > 1)
-              std::cout << "WEIGHT TOO HIGH " << weight << " " << fmod(tstop.t,_mbtime) << std::endl;
-            double rand = randomFlat_.fire();
-            if (weight > rand){
-              stop = tstop;
-              break;
-            }
-          }else{
-            stop = tstop;
-            break;
-          }
-        }
-      }
-    }else{
-      timemap->SetTime(protonPulse_->fire());
-      if (applySurvivalProbability_){
-        while (true){
-          const auto& tstop = stops_.fire();
-          double weight = exp(-tstop.tauNormalized)*survivalProbScaling_;
-          double rand = randomFlat_.fire();
-          if (weight > rand){
-            stop = tstop;
-            break;
+  
+  double RPCGun::MakeEventWeight(art::Ptr<SimParticle> part){ 
+      const PhysicsParams& gc = *GlobalConstantsHandle<PhysicsParams>();
+      double weight = 0.;
+      double tau = part->endProperTime() / gc.getParticleLifetime(part->pdgId());
+     
+      while(part->parent().isNonnull()) { //while particle has a parent which is not null
+        part = part->parent();
+        if ( abs(part->pdgId()) == PDGCode::pi_plus ) { //if pion
+            tau += part->endProperTime() / gc.getParticleLifetime(part->pdgId()); 
           }
-        }
-      }else{
-        const auto& tstop = stops_.fire();
-        stop = tstop;
       }
-    }
-
-    //std::cout << "Found stop " << exp(-stop.tauNormalized) << " " << stop.t << std::endl;
-
-    const CLHEP::Hep3Vector pos(stop.x, stop.y, stop.z);
-
-    const double energy = generateEnergy();
+   
+    weight = exp(-tau);
+    return weight;
+  }
 
-    if(!generateInternalConversion_){
-      output->emplace_back( PDGCode::gamma,
-                            GenId::ExternalRPC,
-                            pos,
-                            CLHEP::HepLorentzVector( randomUnitSphere_.fire(energy), energy),
-                            stop.t );
+  //================================================================
+  void RPCGun::produce(art::Event& event) {
+    auto output{std::make_unique<StageParticleCollection>()};
+    const auto simh = event.getValidHandle<SimParticleCollection>(simsToken_); 
+    const auto pis = stoppedPiMinusList(simh);
+    
+    if(pis.empty()) {
+      throw   cet::exception("BADINPUT")
+        <<"RPCGun::produce(): no suitable stopped pion in the input SimParticleCollection\n";
+    }
 
-      event.put(std::move(output));
-      event.put(std::move(timemap));
-    } else {
+    unsigned int randIn = rand() % pis.size(); 
```

Comment:
I still don't understand if this is using std::rand rather than, say, `randomFlat_.fireint(pis.size())`, which we are setting the seeding for

Reply:
Michael is right, this rand() call is outside the control of the random number seed service.  Please use an engine tied to the seed service.

Reply:
Done